### DISCUSSION TITLE
Custom window manager?

### DISCUSSION THREAD

**Discussion by twert627** [microsoft/wslg#67] [Ideas]
Be able to run another WM than Wayland?

**Comment by spronovo** [comment]
Hi, there are no immediate plan for this. What functionality are you missing the most in the default Weston window manager?

**Reply by nielsdekoeijer** [reply]
I'm sure this has been requested a lot, but personally, I see value in being able to emulate a Linux desktop on windows. This allows for a best-of-both-worlds for those of us who have to run windows at work who would like to use a Linux desktop environment.

**Reply by Apreche** [reply]
> 
> 
> Hi, there are no immediate plan for this. What functionality are you missing the most in the default Weston window manager?

@sprnovo Currently the only things I use with WSL are terminals and vim. You are probably wondering then, why do I need a GUI like wslg at all? It's because managing a bunch of terminals and text editors as individual windows is a pain. In my current configuration I use VcXsrv to run the i3 window manager full screen on one of my two monitors. I would very much like it if wslg could support that, so I would be able to get rid of VcXsrv and also, presumably, experience a great performance benefit from wslg. However, if wslg can not offer this feature, then it's completely useless to me.

As a point of reference, here is a screen capture of the dual monitor scenario I would like to replicate using WSLg, and which currently works almost perfectly using VcXsrv + wsl.

<img width="3840" alt="Screenshot 2021-09-11 112327" src="https://user-images.githubusercontent.com/5847/132952914-0191ef1c-266d-4ff5-bcfa-ea3f0dbc02cd.png">

**Reply by mwarrenus** [reply]
I'd like to use [ch11ng/exwm](https://github.com/ch11ng/exwm).

> EXWM (Emacs X Window Manager) is a full-featured tiling X window manager for Emacs built on top of [XELB](https://github.com/ch11ng/xelb).  It features:  
> - Fully keyboard-driven operations
> - Hybrid layout modes (tiling & stacking)
> - Dynamic workspace support
> - ICCCM/EWMH compliance
> - (Optional) RandR (multi-monitor) support
> - (Optional) Builtin system tray
> - (Optional) Builtin input method

and

> [XELB](https://github.com/ch11ng/xelb) (X protocol Emacs Lisp Binding) is a pure Elisp implementation of X11 protocol based on the XML description files from XCB project. It features an object-oriented API and permits a certain degree of concurrency.

**Reply by hedges333** [reply]
I'm not sure if it's related to choosing a default window manager, but some applications start without minimization buttons, which is a PITA.  See screenshot.  xfce4-terminal doesn't have a minimization button, although gvim does.  This is WSL2.  I removed VcXsrv entirely, since WSL2 handles windowing now on its own.  When I used WSL1 with VcXsrv, this started with a minimization button.  I didn't change anything else.  Do I need to change to a different window manager to get windows like this to start correctly?  Thanks.

![Screenshot 2022-12-22 093000](https://user-images.githubusercontent.com/11917630/209192398-7776c385-4ebc-42cb-a09b-c0b73eb19445.png)

**Comment by Biswa96** [comment]
What window manager do you want to use?

**Reply by Apreche** [reply]
I don't know about the OP, but if you saw my other comment, I would like to use i3.

**Reply by nielsdekoeijer** [reply]
I agree, I3 would be great. For me, it functions as a sort of generalization of tmux that also supports non terminal interfaces.

**Reply by Apreche** [reply]
@nielsdekoeijer That's actually a good point. I suppose I could probably use a full screen Windows terminal with tmux instead of i3 and it wouldn't be that much different for my use case. Would still greatly prefer i3, though. The one significant difference is that with i3 I can have overlapping tabbed panes.

**Reply by charlie-haley** [reply]
If you wanted i3, Sway is an i3-compatible window manager based on Wayland, looks like that will run on top of WSLg. But, from what I've seen, doesn't look very usable

**Reply by mwarrenus** [reply]
I'd like to use [ch11ng/exwm](https://github.com/ch11ng/exwm).

> EXWM (Emacs X Window Manager) is a full-featured tiling X window manager for Emacs built on top of [XELB](https://github.com/ch11ng/xelb).

and

> XELB (X protocol Emacs Lisp Binding) is a pure Elisp implementation of X11 protocol based on the XML description files from XCB project. It features an object-oriented API and permits a certain degree of concurrency.

**Reply by unbrand** [reply]
I vote for openbox. Please. If I could use openbox, I'd ditch this other box I'm using for Linux and use my Windows machine exclusively. Thanks guys for doing such a good job with wsl2!

**Reply by SinclairGurny** [reply]
> If you wanted i3, Sway is an i3-compatible window manager based on Wayland, looks like that will run on top of WSLg. But, from what I've seen, doesn't look very usable

Agreed. I recently tried installing sway. It opened but it did not work properly

**Reply by mid-kid** [reply]
The problem with sway on WSLg is that it doesn't receive all keystrokes and it's a nested compositor, so some applications pop up outside of it. It might genuinely work better to run sway under a windows-based X11 server...

**Comment by davepoltorak** [comment]
Hi Everyone,

Has there been any progress on this particular suggestion? Having recently switched to Windows 11, it would be great to stop dual booting and use WSLg for a performant i3 experience from within Windows.

I've tried VcXsrv and the experience is just a bit too latent to switch away from a bare-metal approach

Thanks for any update!

**Comment by Felix1492** [comment]
I am new to Linux but I have WSL2 Ubuntu 23.10 installed, and then installed I3 and this is the error I get :

I3 >
ERROR: Another window manager is already running (WM_Sn is owned)
ps aux | grep -i wm >
root         534  0.0  0.0   3932  1964 pts/2    S+   20:25   0:00 grep --color=auto -i wm
pkill 534 >
__
ps aux | grep -i wm >
root         540  0.0  0.0   3932  1984 pts/2    S+   20:27   0:00 grep --color=auto -i wm

Because I can't kill it, and it won't stop running, or even sometimes say that it's running I assume that it is a Windows thing.
I have not configured I3 yet because it seems silly to do that when I can't get it running without a windows x-server.
I'm kind of hoping that this question is very simple for you all Because I can't find a solution or even an explanation of what WM_Sn is doing.

Forgive my formatting I didn't know what to do about it. :P

**Comment by dcorking** [comment]
For my use case (tiling) there may be an alternative approach, as [Windows 11's snap feature](https://support.microsoft.com/en-us/windows/snap-your-windows-885a9b1e-a983-a3b1-16cd-c531795e6241) is quite effective for casual use, but doesn't manage my X apps running under WSL. At the risk of sidetracking the OP's feature request: how complex would you guess it might be to add support for the Windows shell's snap UI to wslg?

**Reply by HacknBashe** [reply]
I'm in a similar boat to you. I installed https://github.com/glzr-io/glazewm and it works really well as an i3/sway replacement. The only problem I have with it is it doesn't work with wslg apps.

While being able to run my own wm would be best, glaze is a pretty good replacement for my needs.

**Reply by dcorking** [reply]
Thank you @NickHackford , your idea led me to https://github.com/LGUG2Z/komorebi which I'll try as it appears to [support tiling WSLg windows ](https://github.com/LGUG2Z/komorebi/issues/53)

**Reply by HacknBashe** [reply]
Thanks for the heads up. Komorebi seems to be faster than Glaze but doesn't seem to work as well for me. It seems unable to resize certain windows even after I followed the config tutorial. I also can't get it to handle multiple monitors in a way that is acceptable for me so I will have to stick with Glaze for now.
